Location based search #22

JackCollins91 opened the pull request:
Closes Issue #22  

edgar-tool now supports parameters ```inc_in``` for 'incorporated_in' and ```peo_in``` for 'principal executive offices in.'

both parameters takes a string like "NY" or "NY, OH." Any country code in ```edgar_tool.constants.TEXT_SEARCH_LOCATIONS_MAPPING``` is supported. 

Also, the country names as per TEXT_SEARCH_LOCATIONS_MAPPING.values() are accepted. Whitespace and upper/lower casing should be ignored, and so therefore, "mexico,ny" and "MexicO, Ny" will both convert to "NY,O5" 

Malformed input will result in the dictionary for ```edgar_tool.constants.TEXT_SEARCH_LOCATIONS_MAPPING``` being printed in the console so the user can see what codes they can use. 

I tested the following commands for both correct, incorrect, and unrelated use cases, I have pasted the results of that run in the attached file.  

```
$ poetry run edgar-tool text_search Tsunami Hazards --start_date "2019-06-01" --end_date "2024-01-01" --output "results.csv" --peo_in "NY, OH"
$ poetry run edgar-tool text_search Tsunami Hazards --start_date "2019-06-01" --end_date "2024-01-01" --output "results.csv" --peo_in "NY"
$ poetry run edgar-tool text_search Tsunami Hazards --start_date "2019-06-01" --end_date "2024-01-01" --output "results.csv" --inc_in "NY, OH, O5"
$ poetry run edgar-tool text_search Tsunami Hazards --start_date "2019-06-01" --end_date "2024-01-01" --output "results.csv" --inc_in "NY"
$ poetry run poetry run edgar-tool text_search Tsunami Hazards --start_date "2019-06-01" --end_date "2024-01-01" --output "results.csv" --peo_in "NY, Korea, Democratic People's Republic of, OH, Mexico"
$ poetry run poetry run edgar-tool text_search Tsunami Hazards --start_date "2019-06-01" --end_date "2024-01-01" --output "results.csv" --inc_in "Korea, Democratic People's Republic of"
$ poetry run edgar-tool text_search Tsunami Hazards --output "results.csv" --inc_in "Mexico"
$ poetry run poetry run edgar-tool text_search Tsunami Hazards --start_date "2019-06-01" --end_date "2024-01-01" --output "results.csv" --peo_in "mexico"
$ poetry run poetry run edgar-tool text_search Tsunami Hazards --start_date "2019-06-01" --end_date "2024-01-01" --output "results.csv" --inc_in "kOrEa, demOcratic people's republic of, OH"
$ poetry run edgar-tool text_search Tsunami Hazards --output "results.csv" --peo_in "hello world"
$ poetry run edgar-tool text_search Tsunami Hazards --output "results.csv" --inc_in 123
$ poetry run edgar-tool text_search Tsunami Hazards --start_date "2019-06-01" --end_date "2024-01-01" --output "results.csv" --inc_in "NY" --peo_in "NY, OH"
$ poetry run edgar-tool text_search Tsunami Hazards --start_date "2019-06-01" --end_date "2024-01-01" --output "results.csv"
```

[test cli session.txt](https://github.com/user-attachments/files/15748397/test.cli.session.txt)

GalenReich commented on edgar_tool/utils.py:
```diff
@@ -35,3 +35,50 @@ def safe_get(d: Dict, *keys) -> Any:
+            f'and be one of {TEXT_SEARCH_LOCATIONS_MAPPING}'
+        )
+
+    if isinstance(location_input, tuple):
+        for value in location_input:
+            if value not in TEXT_SEARCH_LOCATIONS_MAPPING:
+                raise ValueError(f"{value} not in {TEXT_SEARCH_LOCATIONS_MAPPING}")
+        location_input = ','.join(location_input)
+
+    elif isinstance(location_input, str):
+        if location_input not in TEXT_SEARCH_LOCATIONS_MAPPING:
+            raise ValueError(f"{location_input} not in {TEXT_SEARCH_LOCATIONS_MAPPING}")
```
It would be great if we could support both the short codes and the long names as in the `--inc_in "Mexico"` example in the issue.

So a user should be able to provide `--inc_in "Mexico"` or `--inc_in "O5"` and achieve the same result.

Some challenges could include the comma and trailing space in some of the entries: `"M4": "Korea, Democratic People's Republic of "`

What do you think about this?

JackCollins91 replied:
Hi @GalenReich I can implement that :) I'll revise the PR in the next few days.

JackCollins91 commented:
Hi @GalenReich, PR is updated per your request. I also updated the initial comment to reflect changes. The attached .txt file shows the results of the latest testing. I added some new test cases which are detailed there also.
